client/httplib/httpclient.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2020 beego
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httplib
+
+import (
+	"net/http"
+)
+
+// Client provides an HTTP client supporting chain call
+type Client struct {
+	Name       string
+	Endpoint   string
+	CommonOpts []BeegoHttpRequestOption
+
+	Setting BeegoHTTPSettings
+	pointer responsePointer
```

Comment:
I think may be we should give up chain invocation if we must keep this field.
I am thinking about using the design proposed by @askuy that we allow users pass pointer like this:
```go
var status int 
var headers http.Header
var httResp http.Response
client.Get(val, path, params, WithStatus(&status), WithHeaders(&headers), WithHttpResponse(httpResp))
```
But I prefer that we could have something like :
```go
type httpClient struct {

}
func (hc *httpClient) Get() HttpResponse, error {
}

type Client struct {
    HTTP *httpClient
}

func (c *Client) Get(value interface, path, opts...) error {
    resp, err := c.Http.Get(path, opts)
    return resp.ResponseForValue(value)
} 
```

Reply:
In second way, if users want to handle the response by themselves, they could do this:
```go
client.Http.Get(xxxx)
```
This is similar to your design.

Reply:
got it.Maybe we can merge two of them? For the since when user want easily get struct value and status code(or header).

Reply:
How do we merge them? could you provide a demo? I could not get your point.

Reply:
Sure.This means that users can use both methods at the same time. For examplet, when user use some method like GetUser, I thought he will perfer  `client.Get(val, path, params, WithStatus(&status)`. This allows the user to judge whether the resource does not exist（403） or an error occurs in the acquisition（5xx）.

In some case, user will perfer to use `client.Http.Get(xxxx)`. Like read file stream or raw reponse body. I think pass point with option and handle original response is not mutex. We can make them coexist。

Reply:
Good, I think this is the best solution. So in this solution we don't need to store the response inside the `req`?

Reply:
BeegoHttpRequestOption will recive a new param to save response. And pointer intstance will create after req created.
```go
// add new param
type BeegoHttpRequestOption func(request *BeegoHTTPRequest, pointer *responsePointer)

// init response pointer and return to the handleResponse method
func (c *Client) customReq(req *BeegoHTTPRequest, opts []BeegoHttpRequestOption) *responsePointer {
	req.Setting(c.Setting)
	opts = append(c.CommonOpts, opts...)
	rp := &responsePointer{}
	for _, o := range opts {
		o(req, rp)
	}
	return rp
}
```

Reply:
How about this:
```go
var resp http.Response
client.Get(xx, xx, xx, WithHttpResponse(&resp))
```
So we don't need to introduce the new structure `responsePointer`.
In your previous design, I think we actually don't need this structure.

Reply:
Is there some way to get `&resp` in `client.Get(xx, xx, xx, WithHttpResponse(&resp))` without any new structure or edit exist structure? :flushed:  I don't know how to solve it

Reply:
Ok, I got your concern.

Let's make this problem simple.
How about this:
1. If user want to handle http response manually, it use `client.Http.Get()`
2. If user don't want to handle http response manually, it use `client.Get()`

Another choice is that we return two value:
```go
func (c *Client) Get(...) (http.Response, error)
```